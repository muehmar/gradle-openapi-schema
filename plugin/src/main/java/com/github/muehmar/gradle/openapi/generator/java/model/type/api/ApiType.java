package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.ParameterizedApiClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeConversion;
import lombok.Value;

@Value
public class ApiType {
  QualifiedClassName className;
  ParameterizedApiClassName parameterizedClassName;
  ToApiTypeConversion toApiTypeConversion;
  FromApiTypeConversion fromApiTypeConversion;

  public static ApiType fromConversion(
      QualifiedClassName className, TypeConversion conversion, PList<JavaType> generics) {
    final ParameterizedApiClassName parameterizedClassName =
        ParameterizedApiClassName.ofClassNameAndGenerics(className, generics);
    final ToApiTypeConversion toApiTypeConversion =
        new ToApiTypeConversion(ConversionMethod.ofString(className, conversion.getToCustomType()));
    final FromApiTypeConversion fromApiTypeConversion =
        new FromApiTypeConversion(
            ConversionMethod.ofString(className, conversion.getFromCustomType()));
    return new ApiType(
        className, parameterizedClassName, toApiTypeConversion, fromApiTypeConversion);
  }
}
