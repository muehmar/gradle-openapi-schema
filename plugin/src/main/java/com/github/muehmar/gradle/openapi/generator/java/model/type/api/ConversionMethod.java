package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class ConversionMethod {
  private final FactoryMethodConversion factoryMethodConversion;
  private final InstanceMethodConversion instanceMethodConversion;

  private ConversionMethod(
      FactoryMethodConversion factoryMethodConversion,
      InstanceMethodConversion instanceMethodConversion) {
    this.factoryMethodConversion = factoryMethodConversion;
    this.instanceMethodConversion = instanceMethodConversion;
  }

  public static ConversionMethod ofString(String conversionMethod) {
    return FactoryMethodConversion.fromString(conversionMethod)
        .map(ConversionMethod::ofFactoryMethod)
        .orElseGet(
            () ->
                ConversionMethod.ofInstanceMethod(
                    InstanceMethodConversion.ofString(conversionMethod)));
  }

  static ConversionMethod ofFactoryMethod(FactoryMethodConversion factoryMethodConversion) {
    return new ConversionMethod(factoryMethodConversion, null);
  }

  static ConversionMethod ofInstanceMethod(InstanceMethodConversion instanceMethodConversion) {
    return new ConversionMethod(null, instanceMethodConversion);
  }

  public <T> T fold(
      Function<FactoryMethodConversion, T> onFactoryMethod,
      Function<InstanceMethodConversion, T> onInstanceMethod) {
    if (factoryMethodConversion != null) {
      return onFactoryMethod.apply(factoryMethodConversion);
    } else {
      return onInstanceMethod.apply(instanceMethodConversion);
    }
  }
}
