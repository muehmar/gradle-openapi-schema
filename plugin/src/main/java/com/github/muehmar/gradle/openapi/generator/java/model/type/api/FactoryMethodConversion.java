package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import java.util.Optional;
import lombok.Value;

@Value
public class FactoryMethodConversion {
  QualifiedClassName className;
  Name methodName;

  public static Optional<FactoryMethodConversion> fromString(
      QualifiedClassName className, String factoryMethodConversion) {
    final String[] parts = factoryMethodConversion.split("#");
    if (parts.length != 2 && factoryMethodConversion.contains("#")) {
      throw createInvalidFormatException(factoryMethodConversion);
    }
    if (parts.length < 2) {
      return Optional.empty();
    }

    try {
      final QualifiedClassName parsedClassName = QualifiedClassName.ofQualifiedClassName(parts[0]);
      final Name methodName = Name.ofString(parts[1]);
      if (not(parsedClassName.getPackageName().isPresent())
          && parsedClassName.getClassName().equals(className.getClassName())) {
        return Optional.of(new FactoryMethodConversion(className, methodName));
      }
      return Optional.of(new FactoryMethodConversion(parsedClassName, methodName));
    } catch (Exception e) {
      throw createInvalidFormatException(factoryMethodConversion);
    }
  }

  private static OpenApiGeneratorException createInvalidFormatException(
      String factoryMethodConversion) {
    return new OpenApiGeneratorException(
        "Invalid configuration for type conversion '%s'. Expected format for static factory methods: 'package.name.ClassName#methodName'.",
        factoryMethodConversion);
  }
}
