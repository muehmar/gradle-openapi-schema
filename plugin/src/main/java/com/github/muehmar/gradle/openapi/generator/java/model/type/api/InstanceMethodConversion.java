package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import lombok.Value;

@Value
public class InstanceMethodConversion {
  Name methodName;

  public static InstanceMethodConversion ofString(String methodName) {
    return new InstanceMethodConversion(Name.ofString(methodName));
  }
}
