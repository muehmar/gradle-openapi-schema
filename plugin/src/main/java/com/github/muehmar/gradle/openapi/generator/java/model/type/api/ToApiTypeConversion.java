package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import java.util.function.Function;
import lombok.Value;

@Value
public class ToApiTypeConversion {
  ConversionMethod conversionMethod;

  public ToApiTypeConversion(ConversionMethod conversionMethod) {
    this.conversionMethod = conversionMethod;
  }

  public <T> T fold(
      Function<FactoryMethodConversion, T> onFactoryMethod,
      Function<InstanceMethodConversion, T> onInstanceMethod) {
    return conversionMethod.fold(onFactoryMethod, onInstanceMethod);
  }
}
