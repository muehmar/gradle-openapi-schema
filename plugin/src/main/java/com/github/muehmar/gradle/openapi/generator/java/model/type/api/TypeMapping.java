package com.github.muehmar.gradle.openapi.generator.java.model.type.api;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.QualifiedClassName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.settings.ClassTypeMapping;
import com.github.muehmar.gradle.openapi.generator.settings.FormatTypeMapping;
import java.util.Optional;
import lombok.Value;

@Value
public class TypeMapping {
  QualifiedClassName className;
  Optional<ApiType> apiType;

  public static TypeMapping fromClassName(QualifiedClassName className) {
    return new TypeMapping(className, Optional.empty());
  }

  public static TypeMapping fromClassMappings(
      QualifiedClassName originalClassName, PList<ClassTypeMapping> classTypeMappings) {
    return fromClassMappings(originalClassName, classTypeMappings, PList.empty());
  }

  public static TypeMapping fromClassMappings(
      QualifiedClassName originalClassName,
      PList<ClassTypeMapping> classTypeMappings,
      PList<JavaType> generics) {
    return classTypeMappings
        .filter(
            classMapping ->
                classMapping.getFromClass().equals(originalClassName.getClassName().asString()))
        .headOption()
        .map(classMapping -> fromClassMapping(originalClassName, classMapping, generics))
        .orElseGet(() -> fromClassName(originalClassName));
  }

  private static TypeMapping fromClassMapping(
      QualifiedClassName originalClassName,
      ClassTypeMapping classTypeMapping,
      PList<JavaType> generics) {
    final QualifiedClassName mappedClassName =
        QualifiedClassName.fromClassTypeMapping(classTypeMapping);

    return classTypeMapping
        .getTypeConversion()
        .map(conversion -> ApiType.fromConversion(mappedClassName, conversion, generics))
        .map(apiType -> new TypeMapping(originalClassName, Optional.of(apiType)))
        .orElseGet(() -> fromClassName(mappedClassName));
  }

  public static TypeMapping fromFormatMappings(
      QualifiedClassName originalClassName,
      String formatString,
      PList<FormatTypeMapping> formatTypeMappings) {
    return fromFormatMappings(originalClassName, formatString, formatTypeMappings, PList.empty());
  }

  public static TypeMapping fromFormatMappings(
      QualifiedClassName originalClassName,
      String formatString,
      PList<FormatTypeMapping> formatTypeMappings,
      PList<JavaType> generics) {
    return formatTypeMappings
        .filter(formatMapping -> formatMapping.getFormatType().equals(formatString))
        .headOption()
        .map(formatMapping -> fromFormatMapping(originalClassName, formatMapping, generics))
        .orElseGet(() -> fromClassName(originalClassName));
  }

  private static TypeMapping fromFormatMapping(
      QualifiedClassName originalClassName,
      FormatTypeMapping formatTypeMapping,
      PList<JavaType> generics) {
    final QualifiedClassName mappedClassName =
        QualifiedClassName.ofQualifiedClassName(formatTypeMapping.getClassType());

    return formatTypeMapping
        .getTypeConversion()
        .map(conversion -> ApiType.fromConversion(mappedClassName, conversion, generics))
        .map(apiType -> new TypeMapping(originalClassName, Optional.of(apiType)))
        .orElseGet(() -> fromClassName(mappedClassName));
  }

  public TypeMapping or(TypeMapping other, QualifiedClassName originalClassName) {
    if (this.apiType.isPresent()) {
      return this;
    } else if (other.getApiType().isPresent()) {
      return other;
    } else if (this.className.equals(originalClassName)) {
      return other;
    } else {
      return this;
    }
  }
}
