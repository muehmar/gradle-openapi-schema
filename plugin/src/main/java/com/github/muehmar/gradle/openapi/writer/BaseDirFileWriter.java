package com.github.muehmar.gradle.openapi.writer;

import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

public class BaseDirFileWriter implements FileWriter {
  private final String outputDir;

  public BaseDirFileWriter(String outputDir) {
    this.outputDir = outputDir;
  }

  @Override
  public void writeFile(GeneratedFile file) {
    try {
      final Path completePath = Path.of(outputDir).resolve(file.getFile());
      completePath.toFile().mkdirs();

      Files.delete(completePath);

      try (final PrintWriter pw =
          new PrintWriter(
              new OutputStreamWriter(
                  Files.newOutputStream(completePath.toFile().toPath()), StandardCharsets.UTF_8))) {

        pw.append(file.getContent());
        pw.flush();
      }
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }
}
