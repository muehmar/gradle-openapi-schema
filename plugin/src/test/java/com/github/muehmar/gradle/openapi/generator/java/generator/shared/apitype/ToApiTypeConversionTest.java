package com.github.muehmar.gradle.openapi.generator.java.generator.shared.apitype;

import static com.github.muehmar.gradle.openapi.generator.java.generator.data.VoidData.noSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.api.ApiType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.api.ApiTypes;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

@SnapshotTest
class ToApiTypeConversionTest {
  private Expect expect;

  @ParameterizedTest
  @EnumSource(ConversionGenerationMode.class)
  @SnapshotName("apiTypeWithFactoryMethods")
  void generate_when_apiTypeWithFactoryMethods_then_matchSnapshot(ConversionGenerationMode mode) {
    final Generator<ApiType, Void> generator =
        ToApiTypeConversion.toApiTypeConversion("field", mode);

    final Writer writer = generator.generate(ApiTypes.userId(), noSettings(), javaWriter());

    expect.scenario(mode.name()).toMatchSnapshot(writerSnapshot(writer));
  }

  @ParameterizedTest
  @EnumSource(ConversionGenerationMode.class)
  @SnapshotName("apiTypeWithInstanceMethods")
  void generate_when_apiTypeWithInstanceMethods_then_matchSnapshot(ConversionGenerationMode mode) {
    final Generator<ApiType, Void> generator =
        ToApiTypeConversion.toApiTypeConversion("field", mode);

    final Writer writer = generator.generate(ApiTypes.counter(), noSettings(), javaWriter());

    expect.scenario(mode.name()).toMatchSnapshot(writerSnapshot(writer));
  }
}
