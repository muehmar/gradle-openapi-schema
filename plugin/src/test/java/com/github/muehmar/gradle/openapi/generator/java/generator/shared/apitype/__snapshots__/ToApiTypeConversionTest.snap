apiTypeWithFactoryMethods[NO_NULL_CHECK]=[
com.github.muehmar.UserId

UserId.fromString(field)
]


apiTypeWithFactoryMethods[NULL_SAFE]=[
com.github.muehmar.UserId

field != null ? UserId.fromString(field) : null
]


apiTypeWithInstanceMethods[NO_NULL_CHECK]=[
.
.
field.toCounter()
]


apiTypeWithInstanceMethods[NULL_SAFE]=[
.
.
field != null ? field.toCounter() : null
]